File: UnrealEngine/Engine/Source/Programs/UnrealGameSync/UnrealGameSyncShared/BuildStep.cs
// Copyright Epic Games, Inc. All Rights Reserved.

using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace UnrealGameSync
{
	public enum BuildStepType
	{
		Compile,
		Cook,
		Other,
	}

	[DebuggerDisplay("{Description}")]
	public class BuildStep
	{
		public const string UniqueIdKey = "UniqueId";

		public Guid UniqueId { get; set; }
		public int OrderIndex { get; set; }
		public string? Description { get; set; }
		public string? StatusText { get; set; }
		public int EstimatedDuration { get; set; }
		public BuildStepType Type { get; set; }
		public string? Target { get; set; }
		public string? Platform { get; set; }
		public string? Configuration { get; set; }
		public string? FileName { get; set; }
		public string? WorkingDir { get; set; }
		public string? Arguments { get; set; }
		public bool UseLogWindow { get; set; }
		public bool NormalSync { get; set; }
		public bool ScheduledSync { get; set; }
		public string? StatusPanelLink { get; set; }
		public List<Guid> Requires { get; } = new List<Guid>();
		public Guid ToolId { get; set; }

		public BuildStep(Guid inUniqueId, int inOrderIndex, string? inDescription, string? inStatusText, int inEstimatedDuration, string? inFileName, string? inArguments, string? inWorkingDir, bool inUseLogWindow)
		{
			UniqueId = inUniqueId;
			OrderIndex = inOrderIndex;
			Description = inDescription;
			StatusText = inStatusText;
			EstimatedDuration = inEstimatedDuration;
			Type = BuildStepType.Other;
			FileName = inFileName;
			Arguments = inArguments;
			WorkingDir = inWorkingDir;
			UseLogWindow = inUseLogWindow;
		}

		public BuildStep(Guid inUniqueId, int inOrderIndex, string? inDescription, string? inStatusText, int inEstimatedDuration, string? inTarget, string? inPlatform, string? inConfiguration, string? inArguments, bool inSyncDefault)
		{
			UniqueId = inUniqueId;
			OrderIndex = inOrderIndex;
			Description = inDescription;
			StatusText = inStatusText;
			EstimatedDuration = inEstimatedDuration;
			Type = BuildStepType.Compile;
			Target = inTarget;
			Platform = inPlatform;
			Configuration = inConfiguration;
			Arguments = inArguments;
			UseLogWindow = true;
			NormalSync = inSyncDefault;
			ScheduledSync = inSyncDefault;
		}

		public BuildStep(ConfigObject obj)
		{
			UniqueId = obj.GetValue(UniqueIdKey, Guid.NewGuid());
			OrderIndex = obj.GetValue("OrderIndex", -1);
			Description = obj.GetValue("Description", "Untitled");
			StatusText = obj.GetValue("StatusText", "Untitled");
			EstimatedDuration = Math.Max(obj.GetValue("EstimatedDuration", 1), 1);

			BuildStepType stepType;
			if (!Enum.TryParse(obj.GetValue("Type", ""), true, out stepType))
			{
				stepType = BuildStepType.Other;
			}
			Type = stepType;

			Target = obj.GetValue("Target");
			Platform = obj.GetValue("Platform");
			Configuration = obj.GetValue("Configuration");
			FileName = obj.GetValue("FileName");
			WorkingDir = obj.GetValue("WorkingDir");
			Arguments = obj.GetValue("Arguments");

			UseLogWindow = obj.GetValue("bUseLogWindow", true);
			NormalSync = obj.GetValue("bNormalSync", true);
			ScheduledSync = obj.GetValue("bScheduledSync", NormalSync);

			StatusPanelLink = obj.GetValue("Link", null);
			if (String.IsNullOrEmpty(StatusPanelLink))
			{
				bool showAsTool;
				if (Boolean.TryParse(obj.GetValue("bShowAsTool", ""), out showAsTool) && showAsTool)
				{
					StatusPanelLink = $"More...|{Description}";
				}
			}

			List<Guid> requires = new List<Guid>();
			foreach (string requireString in obj.GetValue("Requires", String.Empty).Split(';', StringSplitOptions.RemoveEmptyEntries))
			{
				if (Guid.TryParse(requireString, out Guid require))
				{
					requires.Add(require);
				}
			}
			Requires = requires;

			ToolId = obj.GetValue("Tool", Guid.Empty);
		}

		public bool IsValid()
		{
			switch (Type)
			{
				case BuildStepType.Compile:
					return Target != null && Platform != null && Configuration != null;
				case BuildStepType.Cook:
					return FileName != null;
				case BuildStepType.Other:
					return FileName != null;
				default:
					return false;
			}
		}

		public static void MergeBuildStepObjects(Dictionary<Guid, ConfigObject> buildStepObjects, IEnumerable<ConfigObject> modifyObjects)
		{
			foreach (ConfigObject modifyObject in modifyObjects)
			{
				Guid uniqueId;
				if (Guid.TryParse(modifyObject.GetValue(UniqueIdKey, ""), out uniqueId))
				{
					ConfigObject? defaultObject;
					if (buildStepObjects.TryGetValue(uniqueId, out defaultObject))
					{
						modifyObject.SetDefaults(defaultObject);
					}
					buildStepObjects[uniqueId] = modifyObject;
				}
			}
		}

		public ConfigObject ToConfigObject()
		{
			ConfigObject result = new ConfigObject();
			result["UniqueId"] = UniqueId.ToString();
			result["Description"] = Description;
			result["StatusText"] = StatusText;
			result["EstimatedDuration"] = EstimatedDuration.ToString();
			result["Type"] = Type.ToString();
			switch (Type)
			{
				case BuildStepType.Compile:
					result["Target"] = Target;
					result["Platform"] = Platform;
					result["Configuration"] = Configuration;
					result["Arguments"] = Arguments;
					break;
				case BuildStepType.Cook:
					result["FileName"] = FileName;
					break;
				case BuildStepType.Other:
					result["FileName"] = FileName;
					result["WorkingDir"] = WorkingDir;
					result["Arguments"] = Arguments;
					result["bUseLogWindow"] = UseLogWindow.ToString();
					break;
			}
			result["OrderIndex"] = OrderIndex.ToString();
			result["bNormalSync"] = NormalSync.ToString();
			result["bScheduledSync"] = ScheduledSync.ToString();
			if (!String.IsNullOrEmpty(StatusPanelLink))
			{
				result["Link"] = StatusPanelLink;
			}
			if (ToolId != Guid.Empty)
			{
				result["Tool"] = ToolId.ToString();
			}
			return result;
		}

		public ConfigObject? ToConfigObject(ConfigObject? defaultObject)
		{
			ConfigObject result = new ConfigObject();
			result[UniqueIdKey] = UniqueId.ToString();
			result.AddOverrides(ToConfigObject(), defaultObject);
			return (result.Pairs.Count <= 1) ? null : result;
		}
	}
}

File: UnrealEngine/Engine/Source/Programs/UGSCore/BuildStep.cpp
// Copyright Epic Games, Inc. All Rights Reserved.

#include "BuildStep.h"

namespace UGSCore
{

const TCHAR* FBuildStep::UniqueIdKey = TEXT("UniqueId");

bool TryParse(const TCHAR* Text, EBuildStepType& OutType)
{
	if (FCString::Stricmp(Text, TEXT("Compile")) == 0)
	{
		OutType = EBuildStepType::Compile;
		return true;
	}
	if (FCString::Stricmp(Text, TEXT("Cook")) == 0)
	{
		OutType = EBuildStepType::Cook;
		return true;
	}
	if (FCString::Stricmp(Text, TEXT("Other")) == 0)
	{
		OutType = EBuildStepType::Other;
		return true;
	}
	return false;
}

const TCHAR* ToString(EBuildStepType Type)
{
	switch(Type)
	{
	case EBuildStepType::Compile:
		return TEXT("Compile");
	case EBuildStepType::Cook:
		return TEXT("Cook");
	case EBuildStepType::Other:
		return TEXT("Other");
	default:
		return TEXT("<Invalid>");
	}
}

FBuildStep::FBuildStep(const FGuid& InUniqueId, int InOrderIndex, const FString& InDescription, const FString& InStatusText, int InEstimatedDuration, const FString& InTarget, const FString& InPlatform, const FString& InConfiguration, const FString& InArguments, bool bInSyncDefault)
	: UniqueId(InUniqueId)
	, OrderIndex(InOrderIndex)
	, Description(InDescription)
	, StatusText(InStatusText)
	, EstimatedDuration(InEstimatedDuration)
	, Type(EBuildStepType::Compile)
	, Target(InTarget)
	, Platform(InPlatform)
	, Configuration(InConfiguration)
	, Arguments(InArguments)
	, bUseLogWindow(false)
	, bNormalSync(bInSyncDefault)
	, bScheduledSync(bInSyncDefault)
	, bShowAsTool(false)
{
}

FBuildStep::FBuildStep(const FCustomConfigObject& Object)
{
	if(!Object.TryGetValue(UniqueIdKey, UniqueId))
	{
		UniqueId = FGuid::NewGuid();
	}
	if(!Object.TryGetValue(TEXT("OrderIndex"), OrderIndex))
	{
		OrderIndex = -1;
	}

	Description = Object.GetValueOrDefault(TEXT("Description"), TEXT("Untitled"));
	StatusText = Object.GetValueOrDefault(TEXT("StatusText"), TEXT("Untitled"));

	if(!Object.TryGetValue(TEXT("EstimatedDuration"), EstimatedDuration) || EstimatedDuration < 1)
	{
		EstimatedDuration = 1;
	}

	if(!TryParse(Object.GetValueOrDefault(TEXT("Type"), TEXT("")), Type))
	{
		Type = EBuildStepType::Other;
	}

	Object.TryGetValue(TEXT("Target"), Target);
	Object.TryGetValue(TEXT("Platform"), Platform);
	Object.TryGetValue(TEXT("Configuration"), Configuration);
	Object.TryGetValue(TEXT("FileName"), FileName);
	Object.TryGetValue(TEXT("Arguments"), Arguments);

	bUseLogWindow = Object.GetValueOrDefault(TEXT("bUseLogWindow"), true);
	bNormalSync = Object.GetValueOrDefault(TEXT("bNormalSync"), true);
	bScheduledSync = Object.GetValueOrDefault(TEXT("bScheduledSync"), true);
	bShowAsTool = Object.GetValueOrDefault(TEXT("bShowAsTool"), false);
}

void FBuildStep::MergeBuildStepObjects(TMap<FGuid, FCustomConfigObject>& BuildStepObjects, const TArray<FCustomConfigObject>& ModifyObjects)
{
	for(const FCustomConfigObject& ModifyObject : ModifyObjects)
	{
		FGuid UniqueId;
		if(ModifyObject.TryGetValue(UniqueIdKey, UniqueId))
		{
			FCustomConfigObject CombinedObject = ModifyObject;

			FCustomConfigObject* DefaultObject = BuildStepObjects.Find(UniqueId);
			if(DefaultObject != nullptr)
			{
				CombinedObject.SetDefaults(*DefaultObject);
			}

			BuildStepObjects[UniqueId] = CombinedObject;
		}
	}
}

FCustomConfigObject FBuildStep::ToConfigObject() const
{
	FCustomConfigObject Result;
	Result.SetValue(TEXT("UniqueId"), UniqueId);
	Result.SetValue(TEXT("Description"), *Description);
	Result.SetValue(TEXT("StatusText"), *StatusText);
	Result.SetValue(TEXT("EstimatedDuration"), EstimatedDuration);
	Result.SetValue(TEXT("Type"), ToString(Type));
	switch(Type)
	{
	case EBuildStepType::Compile:
		Result.SetValue(TEXT("Target"), *Target);
		Result.SetValue(TEXT("Platform"), *Platform);
		Result.SetValue(TEXT("Configuration"), *Configuration);
		Result.SetValue(TEXT("Arguments"), *Arguments);
		break;
	case EBuildStepType::Cook:
		Result.SetValue(TEXT("FileName"), *FileName);
		break;
	case EBuildStepType::Other:
		Result.SetValue(TEXT("FileName"), *FileName);
		Result.SetValue(TEXT("Arguments"), *Arguments);
		Result.SetValue(TEXT("bUseLogWindow"), bUseLogWindow);
		break;
	}
	Result.SetValue(TEXT("OrderIndex"), OrderIndex);
	Result.SetValue(TEXT("bNormalSync"), bNormalSync);
	Result.SetValue(TEXT("bScheduledSync"), bScheduledSync);
	Result.SetValue(TEXT("bShowAsTool"), bShowAsTool);
	return Result;
}

FCustomConfigObject FBuildStep::ToConfigObject(const FCustomConfigObject* DefaultObject) const
{
	FCustomConfigObject Result;
	Result.SetValue(UniqueIdKey, UniqueId);
	Result.AddOverrides(ToConfigObject(), DefaultObject);
	return Result;
}

} // namespace UGSCore
